bottlecap/src/appsec/processor/context.rs

```diff
@@ -24,25 +30,134 @@ pub struct Context {
+    }
+
+    /// Marks the response of this request as having been seen.
+    pub(super) async fn set_response_seen(&mut self) {
+        if self.response_seen {
+            return;
+        }
+
+        if let Some((mut trace, sender, args)) = self.on_response_seen.take() {
+            if let Some(span) = Processor::service_entry_span_mut(&mut trace) {
+                // Debug-sanity check that we're holding the right trace.
+                debug_assert_eq!(
+                    span.meta.get("request_id").map(String::as_str),
+                    Some(self.rid.as_str())
+                );
+                debug!(
+                    "aap: processing trace for request {} now that the response was seen",
+                    self.rid
+                );
+                self.process_span(span);
+            }
+
+            debug!("aap: flushing out trace for request {}", self.rid);
+            match sender
+                .send_processed_traces(
+                    args.config,
+                    args.tags_provider,
+                    tracer_header_tags::TracerHeaderTags {
+                        lang: &args.tracer_header_tags_lang,
+                        lang_version: &args.tracer_header_tags_lang_version,
+                        lang_interpreter: &args.tracer_header_tags_lang_interpreter,
+                        lang_vendor: &args.tracer_header_tags_lang_vendor,
+                        tracer_version: &args.tracer_header_tags_tracer_version,
+                        container_id: &args.tracer_header_tags_container_id,
+                        client_computed_top_level: args
+                            .tracer_header_tags_client_computed_top_level,
+                        client_computed_stats: args.tracer_header_tags_client_computed_stats,
+                        dropped_p0_traces: args.tracer_header_tags_dropped_p0_traces,
+                        dropped_p0_spans: args.tracer_header_tags_dropped_p0_spans,
+                    },
+                    vec![trace],
+                    args.body_size,
+                    args.span_pointers,
+                )
+                .await
+            {
+                Ok(()) => debug!("aap: successfully sent trace to aggregator buffer"),
+                Err(e) => warn!("aap: failed to send trace to aggregator buffer: {e}"),
+            }
+        }
+
+        debug!(
+            "aap: marking security context for {} as having seen the response...",
+            self.rid
+        );
+        self.response_seen = true;
+    }
+
+    /// Holds a trace for future processing once the response is seen.
+    ///
+    /// # Panics
+    /// If called after [`Self::is_pending_response`] returns `false`.
+    pub(crate) fn hold_trace(
+        &mut self,
+        trace: Vec<Span>,
+        sender: SendingTraceProcessor,
+        args: HoldArguments,
+    ) {
+        if !self.is_pending_response() {
+            unreachable!("Context::hold_trace called after response was seen!");
```

Comment:
Maybe we don't want to panic here and instead log and return? How would this look if it panics?

Just ensuring we don't crash the Extension, which in turn would crash the runtime

Reply:
Fair. I guess the `unreachable!` here means it's not supposed to be happening, but that can't be statically guaranteed (or else there would not be this `unreachable!`). I can make it a log unless the extension is built with `debug-assertions`.